File: LearnLink.Shared/Responses/DataPage.cs
namespace LearnLink.Shared.Responses
{
    public class DataPageHeader
    {
        private int pageSize = 1;
        public int PageNumber { get; init; } = 1;
        public int PageSize
        {
            get => pageSize;
            init => pageSize = value == 0 ? 1 : value;
        }

        public DataPageHeader() { }

        public DataPageHeader(int pageNumber, int pageSize)
        {
            PageNumber = pageNumber;
            PageSize = pageSize;
        }
    }
    public class DataPage<T> : DataPageHeader
    {
        public int ItemsCount { get; init; }
        public int PageCount => (int)Math.Ceiling((decimal)ItemsCount / PageSize);

        public T? Values { get; set; }
    }
}

File: frontend/src/models/dataPage.ts
export class DataPageHeader {
    pageSize: number;
    pageNumber: number;

    constructor(pageSize: number, pageNumber: number) {
        this.pageSize = pageSize;
        this.pageNumber = pageNumber;
    }
}

export class DataPage extends DataPageHeader {
    itemsCount: number;
    pageCount: number;

    constructor(pageSize: number, pageNumber: number, itemsCount: number, pageCount: number) {
        super(pageSize, pageNumber);
        this.itemsCount = itemsCount;
        this.pageCount = pageCount;
    }
}

export class ValueDataPage<T> extends DataPage {
    values?: T;

    constructor(pageSize: number, pageNumber: number, itemsCount: number, pageCount: number, values?: T) {
        super(pageSize, pageNumber, itemsCount, pageCount);
        this.values = values;
    }
}